test/e2e_node/node_shutdown_linux_test.go

```diff
@@ -0,0 +1,202 @@
+			for _, pod := range list.Items {
+				if kubelettypes.IsCriticalPod(&pod) {
+					framework.ExpectEqual(
+						pod.Status.Phase == v1.PodRunning,
+						true,
+						"critical pod should not be shutdown",
+					)
+				} else {
+					framework.ExpectEqual(
+						pod.Status.Phase == v1.PodFailed && pod.Status.Reason == "Shutdown",
+						true,
+						"pod should be shutdown",
+					)
+				}
+			}
+
+			time.Sleep(6 * time.Second)
+			// All pod should be shutdown
+			list, err = f.PodClient().List(context.TODO(), metav1.ListOptions{})
+			framework.ExpectNoError(err)
+			for _, pod := range list.Items {
+				framework.ExpectEqual(
+					pod.Status.Phase == v1.PodFailed && pod.Status.Reason == "Shutdown",
+					true,
+					"pod should be shutdown",
+				)
+			}
+		})
+
+		ginkgo.It("Interrupt shutdown", func() {
+			err := emitSignalPrepareForShutdown(true)
+			framework.ExpectNoError(err)
+			for i := 0; ; i++ {
+				isReady := getNodeReadyStatus(f)
+				if !isReady {
+					break
+				}
+				if i == 2 {
+					framework.ExpectEqual(
+						isReady,
+						false,
+						"node did not become shutdown as expected",
+					)
+				} else {
+					time.Sleep(time.Second)
+				}
+			}
+
+			err = emitSignalPrepareForShutdown(false)
+			framework.ExpectNoError(err)
+			for i := 0; ; i++ {
+				isReady := getNodeReadyStatus(f)
+				if isReady {
+					break
+				}
+				if i == 10 {
+					framework.ExpectEqual(
+						isReady,
+						true,
+						"node did not recover as expected",
+					)
+				} else {
+					time.Sleep(time.Second)
+				}
+			}
+		})
+	})
+})
+
+func getGracePeriodOverrideTestPod(name string, node string, gracePeriod int64, critical bool) *v1.Pod {
+	pod := &v1.Pod{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Pod",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: name,
+		},
+		Spec: v1.PodSpec{
+			Containers: []v1.Container{
+				{
+					Name:    name,
+					Image:   busyboxImage,
+					Command: []string{"sh", "-c"},
+					Args: []string{`
+_term() {
+	echo "Caught SIGTERM signal!"
+	sleep 300
+}
+trap _term SIGTERM 
+sleep 300
+`},
+				},
+			},
+			TerminationGracePeriodSeconds: &gracePeriod,
+			NodeName:                      node,
+		},
+	}
+	if critical {
+		pod.ObjectMeta.Annotations = map[string]string{
+			kubelettypes.ConfigSourceAnnotationKey: kubelettypes.FileSource,
+		}
+		pod.Spec.PriorityClassName = scheduling.SystemNodeCritical
+
+		framework.ExpectEqual(kubelettypes.IsCriticalPod(pod), true, "pod should be a critical pod")
+	} else {
+		framework.ExpectEqual(kubelettypes.IsCriticalPod(pod), false, "pod should not be a critical pod")
+	}
+	return pod
+}
+
+func emitSignalPrepareForShutdown(b bool) error {
+	cmd := "gdbus emit --system --object-path /org/freedesktop/login1 --signal org.freedesktop.login1.Manager.PrepareForShutdown " + strconv.FormatBool(b)
+	_, err := runCommand("sh", "-c", cmd)
+	return err
+}
+
+func getNodeReadyStatus(f *framework.Framework) bool {
```

Comment:
you should be able to use getNode which already does this:

https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/e2e_node_suite_test.go#L314

I think all you need is something like this:
https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/e2e_node_suite_test.go#L273-L280

(or maybe above when you check the node status you can use `waitForNodeReady` directly and remove this function?

Reply:
`getNode` may not be used because it is not compatible with `framework.Framework`

'waitForNodeReady' is also not needed. this case not only requires waiting for node ready but also waiting for node not ready.